EightfoldAI/octuple
# feat: support tree shaking

## SUMMARY:
Adds tree-shaking support for apps using Octuple. Core changes:
1. Add explicit `sideEffects` to `package.json` for CJS tree shaking.
2. Switch from Webpack to Rollup to support exporting ESM. 

For the Webpack -> Rollup change, it _may_ be possible to export ESM from webpack, but most sources I found said exporting ESM from webpack was [not supported](https://github.com/webpack/webpack/issues/2933). Additionally, after examining quite a number of specifically react component libraries, it seemed like Rollup was the most common tool for this job.

> Testing on Eightfold for the `careerHub` entry point only yielded a `-20.9%` decrease in the bundle size for the octuple chunk compared to v2.51.1

Here is the bundle analysis diff in Blindfold.ai after following the steps outlined in the Test Plan
### Before
<img width="100%" alt="Before" src="https://github.com/EightfoldAI/octuple/assets/87032295/85ba4bed-a7e2-4a8f-99f3-eb771fb8d322">

### After
<img width="100%" alt="After" src="https://github.com/EightfoldAI/octuple/assets/87032295/90acb341-f3f4-4101-9153-e253b8920bce">

## GITHUB ISSUE (Open Source Contributors)
NA

## JIRA TASK (Eightfold Employees Only):
ENG-55139

## CHANGE TYPE:


## TEST COVERAGE:


## TEST PLAN:
> This test plan relies on running Blindfold locally, but can be done in isolation if desired

1. Clone the https://github.com/EightfoldAI/blindfold repo locally
2. In the root of the repo, add the following `.env` file. (If you have one already, just add `ANALYZE_BUNDLES=1`
```
ANALYZE_BUNDLES=1
BACKEND_BASE_URL="https://app.eightfold.ai"
NEXT_PUBLIC_HOST_URL="https://localhost:8000"
NEXTAUTH_URL="https://localhost:8000"
```
2. Run `git checkout gclark/octuple-bundle-test` to pull [these changes](https://github.com/EightfoldAI/blindfold/pull/453/files).
3. Install deps with `pnpm install`. (Run `npm install -g pnpm` if you do not have pnpm installed)
4. Run `pnpm build --filter nextjs` to build the Blindfold app.
5. Once complete open the file located in `apps/nextjs/.next/analyze/client.html` in your web browser.
    - [x] The bundle analysis does not show the entire `octuple.js` bundle present.
6. Run `pnpm start --filter nextjs` and go to http://localhost:8000/octuple
    - [x] Verify the Octuple primary button is present on the page and styled correctly.
7. Remove the patch with `pnpm patch-remove @eightfold.ai/octuple`
8. Run `pnpm build --filter nextjs` again.
9. Once complete open the file located in `apps/nextjs/.next/analyze/client.html` in your web browser.
    - [x] The bundle analysis **does** show the entire `octuple.js` bundle present.
